addon-resizer/nanny/kubernetes_client.go

```diff
@@ -83,54 +85,70 @@ func hasEqualValues(a string, b *string) bool {
 	return b != nil && a == *b
 }
 
-func (k *kubernetesClient) countNodesThroughMetrics() (uint64, error) {
-	// Similarly as for listing nodes, permissions for /metrics endpoint are needed.
-	// Other than that, endpoint is visible from everywhere.
-	reader, err := k.clientset.CoreV1().RESTClient().Get().RequestURI("/metrics").Stream(context.Background())
-	if err != nil {
-		return 0, err
+func extractMetricValueForNodeCount(mf dto.MetricFamily) (uint64, error) {
+	for _, metric := range mf.Metric {
+		hasLabel := false
+		for _, label := range metric.Label {
+			if hasEqualValues(resourceLabel, label.Name) && hasEqualValues(nodeResourceName, label.Value) {
+				hasLabel = true
+				break
+			}
+		}
+		if !hasLabel {
+			continue
+		}
+		if metric.Gauge == nil || metric.Gauge.Value == nil {
+			continue
+		}
+		if *metric.Gauge.Value < 0 {
+			return 0, fmt.Errorf("metric unknown")
+		}
+		value := uint64(*metric.Gauge.Value)
+		return value, nil
 	}
+	return 0, fmt.Errorf("no valid metric values")
+}
 
-	decoder := expfmt.NewDecoder(reader, expfmt.FmtText)
-
+func getNodeCountFromDecoder(decoder expfmt.Decoder) (uint64, error) {
 	var mf dto.MetricFamily
+	var fallbackNodeCountMetricValue uint64
+	var fallbackMetricError error
+	useNodeCountFromFallbackMetric := false
+
 	for {
 		if err := decoder.Decode(&mf); err != nil {
 			if err == io.EOF {
 				break
 			}
 			return 0, fmt.Errorf("decoding error: %v", err)
 		}
-
-		if !hasEqualValues(objectCountMetricName, mf.Name) {
-			continue
+		if hasEqualValues(objectCountMetricName, mf.Name) {
+			// Preferred metric is present - return immediately
+			return extractMetricValueForNodeCount(mf)
```

Comment:
Should we fall back if extractMetricValueForNodeCount returns error?

Reply:
+1

Reply:
I thought it wouldn't be worth it (since we only expect both metrics in 1.21 and it makes code more complicated for all versions)

Done.